src/Orleans/Utils/TimerWheel.cs

```diff
@@ -0,0 +1,128 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans.Runtime
+{
+    // Removes the need in timer per item, optimized for singlethreaded consumption. 
+    internal class TimerWheel<T> : IDisposable where T : ITimebound
+    {
+        private const int MaxItemsPerLock = 10;
+        private readonly Queue<T> _entries = new Queue<T>();
```

Comment:
One negative caveat of your impl. is that in the fast path case, you will be holding  to the element in this queue for full 30 sec, while in old impl. we disposed timer immediately upon succesfull reply. Therefore, you will be holding callback datas for 30 sec, which can increase memory footprint.
The solution is either to support dispose/unregister, or store weak reference to T.